text/0000-standard-lazy-types.md

```diff
@@ -0,0 +1,416 @@
+
+    // use `Logger::global()` from now on
+}
+```
+
+## Why `OnceCell` as a primitive?
+
+It is possible to imagine a type, slightly more general than `OnceCell`:
+
+```rust
+struct OnceFlipCell<U, V> { ... }
+
+impl<U, V> OnceFlipCell<U, V> {
+    const fn new(initial_value: U) -> OnceFlipCell<U, V>;
+
+    fn get_or_init<F: FnOnce(U) -> V>(&self, f: F) -> &V;
+}
+
+type OnceCell<T> = OnceFlipCell<(), T>;
+```
+
+That is, we can store some initial state in the cell and consume it during initialization.
+In practice, such flexibility seems to be rarely required.
+Even if we add a type, similar to `OnceFlipCell`, having a dedicated `OnceCell` (which *could* be implemented on top of `OnceFlipCell`) type simplifies common use-case.
+
+## Poisoning
+
+As a cell can be empty or fully initialized, the proposed API does not use poisoning.
+If an initialization function panics, the cell remains uninitialized.
+An alternative would be to add poisoning, which will make all subsequent `get` calls to panic.
+
+Similarly, because `OnceCell` provides strong exception safety guarantee, it implements `UnwindSafe`:
+
+```rust
+impl<T: UnwindSafe>                    UnwindSafe for OnceCell<T> {}
+impl<T: UnwindSafe + RefUnwindSafe> RefUnwindSafe for OnceCell<T> {}
+```
+
+## Default type parameter on `Lazy`
+
+`Lazy` is defined with default type parameter.
+
+```rust
+pub struct Lazy<T, F = fn() -> T> { ... }
+```
+
+This is important to make using `Lazy` in static contexts convenient.
+Without this default, the user would have to type `T` type twice:
+
+```rust
+static GLOBAL_DATA: Lazy<Mutex<HashMap<i32, String>>, fn() -> Mutex<HashMap<i32, String>>
+    = Lazy::new(|| ... );
+```
+
+If we allow type inference in statics, this could be shortened to
+
+```rust
+static GLOBAL_DATA: Lazy<Mutex<HashMap<i32, String>>, _>
+    = Lazy::new(|| ... );
+```
+
+There are two drawbacks of using fn pointer type:
+
+* fn pointers are not ZSTs, so we waste one pointer per static lazy value.
+  Lazy locals will generally rely on type-inference and will use more specific closure type.
+* Specifying type for local lazy value might be tricky: `let x: Lazy<i32> = Lazy::new(|| closed_over_var)` fails with type error, the correct syntax is `let x: Lazy<i32, _> = Lazy::new(|| closed_over_var)`.
+
+## Only thread-safe flavor
+
+It is possible to add only `sync` version of the types, as they are the most useful.
+However, this will be against zero cost abstractions spirit.
+Additionally, non thread-safe version is required to replace `thread_local!` macro without imposing synchronization.
+
+# Prior art
+[prior-art]: #prior-art
+
+The primary bit of prior art here is the [`once_cell`] library, which itself draws on multiple sources:
+
+* [double-checked-cell](https://crates.io/crates/double-checked-cell)
+* [lazy-init](https://crates.io/crates/lazy-init)
+* [lazycell](https://crates.io/crates/lazycell)
+* [mitochondria](https://crates.io/crates/mitochondria)
```

Comment:
So it seems worth discussing why this design and not the ones in the other crates?

Reply:
All significant design alternatives I am aware of are discussed in the RFC. As far as I know, each of this crates provides a subset of once_cell \cup alternatives API, as far as lazy values are concerned.